tools/topology/sof-tgl-max98357a-rt5682.m4

```diff
@@ -282,6 +299,12 @@ ifelse(
 		SSP_CLOCK(fsync, 48000, codec_slave),
 		SSP_TDM(4, 25, 3, 15),
 		SSP_CONFIG_DATA(SSP, SPK_SSP_INDEX, 24)))',
+	CODEC, `MAX98390', `
+	SSP_CONFIG(DSP_B, SSP_CLOCK(mclk, 19200000, codec_mclk_in),
+		SSP_CLOCK(bclk, 6144000, codec_slave),
+		SSP_CLOCK(fsync, 48000, codec_slave),
+		SSP_TDM(4, 32, 3, 15),
```

Comment:
I am still in the dark on the configurations. On capture, the tweeter slots would be empty in the 2speaker case, so how would userspace/CRAS know which channels to use?

Do you have something like a device-specific UCM file that provides a channel map to extract the relevant channels?

You should really explain further in the commit message what the end-to-end configuration is, or assumed to be.

Reply:
yes, CRAS has it in UCM that provides the channel map selection.

CaptureChannelMap "0 1 -1 -1 -1 -1 -1 -1 -1 -1 -1"
or
CaptureChannelMap "0 1 2 3 -1 -1 -1 -1 -1 -1 -1"

I am trying to explain more details in commit message.  Feel free to comment if any better description.

Board A with codec max98390 uses 2 in 4 channels mapping for playback and capture.  
Chan 0 = Left
Chan 1 = Right
&
2 codecs DAI selection
'Left DAI Sel Mux' Left"
'Right DAI Sel Mux' Right"

CRAS UCM defines the CaptureChannelsMap selection.
CaptureChannelMap "0 1 -1 -1 -1 -1 -1 -1 -1 -1 -1"

Board B with codec max98390 uses 4 channels mapping for playback and capture.
Chan 0 = Left
Chan 1 = Right
Chan 2 = Tweeter Left
Chan 3 = Tweeter Right
&
4 codecs DAI selection
'Left DAI Sel Mux' Left"
'Right DAI Sel Mux' Right"
'Tweeter Left DAI Sel Mux' Left"
'Tweeter Right DAI Sel Mux' Right"

CRAS defines the CaptureChannelsMap selection.
CaptureChannelMap "0 1 2 3 -1 -1 -1 -1 -1 -1 -1"

Reply:
Suggested edits:

````
This patch adds support for three configurations:
SSP1 connects max98390 2/4 speakers
SSP2 connects max98390 2 speakers

The SSP TDM configuration uses 4 slots for playback and
4 slots for the echo reference capture - regardless of the number of speakers.

UCM files in userspace specify which channels needs to be used on the specific platform. There is no information reported by the topology/firmware related to valid channels.
````